### ISSUE TITLE
Fast refresh not working properly due to page reload when compiling middleware

### ISSUE THREAD

**Issue by timfuhrmann** [vercel/next.js#31827]
### What version of Next.js are you using?

12.0.4

### What version of Node.js are you using?

16.13.0

### What browser are you using?

Chrome, Safari, Edge

### What operating system are you using?

macOS

### How are you deploying your application?

next dev

### Describe the Bug

Since I don't know the technological background or necessity of the page reload for `middleware`, I'm not sure if that's something that you have to live with when working with `middleware` - but any change, like edits made to react components, often trigger a page reload due to the `middleware` being compiled:

> wait  - compiling /_middleware (middleware only)...

This becomes especially frustrating when working with `getStaticProps`.

### Expected Behavior

No page reload when applying changes ("Fast Refresh").

### To Reproduce

Basic `middleware` implementation.

**Comment by exneval** [comment]
I was able to reproduce this, Developer Experience when using middleware is bad, the fast refresh won't work properly, hope this can be get further investigation

**Comment by dlindenkreuz** [comment]
Possibly related: #30791

**Comment by iksent** [comment]
I thought https://github.com/vercel/next.js/pull/31548 would fix this issue, but I am experiencing the same at the latest v.12.0.7

**Comment by havran** [comment]
Same problem here.

**Comment by balazsorban44** [comment]
This has been fixed, please upgrade `npm i next@latest`.

**Comment by iksent** [comment]
@balazsorban44, which version resolves this issue? `next@latest` is v12.0.7, and it's not fixed there.

**Comment by balazsorban44** [comment]
:thinking: I just tried it locally. Maybe I'm missing something?

**Comment by balazsorban44** [comment]
Sorry. Could you try `canary`?

**Comment by timfuhrmann** [comment]
Just tried `canary`, seems to be fixed 🚀

**Comment by balazsorban44** [comment]
If anyone else still having issues, please open a separate bug report with a reproduction.

**Comment by timfuhrmann** [comment]
Sorry, just tried `canary` again and it doesn't seem to be fixed, there still is a reload which I didn't realize because I worked on a project without `getStaticProps`.

**Comment by timfuhrmann** [comment]
To reproduce, add some JSX and save your project, repeat 4 - 5 times and you should get a reload.

**Comment by balazsorban44** [comment]
Yeah, still no problem on my side. :eyes:

Could you link to your repository? Create a minimal reproduction?

**Comment by timfuhrmann** [comment]
I tried to create a minimal reproduction, but couldn't make the issue occur again yet.

After installing `canary` to the [following repository](https://github.com/timfuhrmann/streaming-platform/tree/middleware), it just seems to take longer for the reload to occur though, but the issue somehow is still present. There's only one `middleware` on app level. To clone and run you would have to create a ["The Movie Database" API key](https://developers.themoviedb.org/3) though. If that isn't an option for you, I could try to work on the minimal reproduction again.

**Comment by balazsorban44** [comment]
I don't have a TMDB API key, but for the sake of pinning down where the problem might be, it is favorable if you could provide an isolated/minimal reproduction anyway.

**Comment by henriqemalheiros** [comment]
Hey! I was facing the same issue until I found the problem: something that is imported in the middleware and somewhere in the app. The import could be anything, including an external dependency.  Here is the minimal reproduction: [henriqemalheiros/middleware-fast-refresh](https://github.com/henriqemalheiros/middleware-fast-refresh).

To reproduce it, though, you need to wait ~20 seconds after the page is fully loaded or the last change was made. If you make a change before the ~20 seconds interval, the fast refresh works as expected. After that, any change to any part of the app triggers a full refresh.

There is a warning logged to the console, but it's to fast to actually read it. In `next@canary`, the same warning is shown by the app before the full refresh so we can read it:

> About to perform a full refresh
>
> Fast Refresh will perform a full reload when you edit a file that is imported by modules outside of the React rendering tree. It is also possible the parent component of the component you edited is a class component, which disables Fast Refresh. Fast Refresh requires at least one parent function component in your React tree.
> 
> You can find more information in the related Webpack error below:
> 
> ```
> Cannot read properties of undefined (reading 'call')
> ```
> 
> You can read more about Fast Refresh in [our documentation](https://nextjs.org/docs/basic-features/fast-refresh#how-it-works).

**Comment by switz** [comment]
@henriqemalheiros nice work! this has been really hurting my productivity, thanks for coming up with a reproduction.

**Comment by lao9s** [comment]
Hi,
I have the same issue. To reproduce it: import a function in react component then the same function import in your _middleware.js. Then, if you have running dev server, stop it and start it: `npm run dev`, whait ~15 secconds then try to do changes on your component, will perform full reload.

As a temporary solution is to have different functions with the same logic and then import them separately into both files(your react component and _middleware.js)

Does anyone know how to solve this to use the same function in both places?

**Comment by switz** [comment]
This is pretty frustrating and severely hindering the development experience. Every time I update code it does a full refresh to a blank page and I have to refresh twice to get it to render.

I have imports in my _middleware – there's not really much I can do to get around that. Can I disable re-compilation on a particular nextjs path? That file isn't changing at all, it should be compiled once until any changes are made.

**Comment by tomduncalf** [comment]
Also seeing frequent reloads when working due to the middleware recompiling.

With `canary` it is perhaps a little less frequent, but instead of just a reload I first get this dialog - to be honest, it was better before, as I didn't have to click a button to reload also.

![image](https://user-images.githubusercontent.com/5458070/149021552-169e7a01-8fc7-4cc7-903c-5067bd4a42bd.png)

I am pretty sure this erromessage is referring to the middleware file, as I am doing/undoing the same change to a styled component over and over which works like 3 in 4 times, and does a full reload 1 in 4. I see `wait  - compiling /_middleware...` in the console despite not changing the middleware.

My middleware is just the basic auth example and is totally self contained, except for importing TS types: `import { NextRequest, NextResponse } from "next/server";` so it's not related to the shared imports some others have mentioned.

I can confirm that removing `middleware.ts` fixes this, but doing an early return from the middleware in dev mode does not help.

**Comment by switz** [comment]
After upgrading to canary I see the dialog as well, but with a different error message. I didn't see anything related to on-demand-entries before upgrading to canary. Hoping this is helpful:

<img width="953" alt="image" src="https://user-images.githubusercontent.com/990166/149123895-a74958c8-c81b-4789-90ab-ba9fad6ba73c.png">

**Comment by tomduncalf** [comment]
@WoutGroenendijk you must still have `import { NextRequest, NextResponse } from "next/server";` though, right? For me this happens just with that as the only import

**Comment by WoutGroenendijk** [comment]
@tomduncalf Yes I still got `import { NextRequest, NextResponse } from "next/server";` I just moved my custom imports of path's and constants inside my `middleware` file, after that it started to work fine.

**Comment by henriqemalheiros** [comment]
> @WoutGroenendijk you must still have `import { NextRequest, NextResponse } from "next/server";` though, right? For me this happens just with that as the only import

@tomduncalf is there any chance you are importing `NextRequest` or `NextResponse` in other parts of the app, like an API route? I didn't check for imports in places other than React components and I don't use TypeScript, but it makes sense for the problem to occur in your use case as well.

---

> @tomduncalf Yes I still got `import { NextRequest, NextResponse } from "next/server";` I just moved my custom imports of path's and constants inside my `middleware` file, after that it started to work fine.

@WoutGroenendijk that's what I did as well:

```
import { API_URL } from '../lib/constants'
```

became

```
const API_URL = process.env.NEXT_PUBLIC_API_URL ?? ''
```

It solved the problem, but that makes the app harder to maintain and it is a change impossible to be made when importing other things, like a dependency (`lodash`, for example).

**Comment by tomduncalf** [comment]
Good question @henriqemalheiros, I’m not sure why I’d be importing them elsewhere but it’s possible, I’ll check. Hopefully this issue can be resolved soon anyway, feels like going back to the dark ages with full page refresh 😂

**Comment by tomduncalf** [comment]
No, I am not importing `NextRequest, NextResponse` from elsewhere. Slightly different error on 12.0.8

![image](https://user-images.githubusercontent.com/5458070/149622064-db266ffc-796d-4e82-8e79-2ac7787f3563.png)

**Comment by smo043** [comment]
Removing custom imports solved the problem, but looking for the final solution.

**Comment by bmichotte** [comment]
Also, worth adding this information here (I guess) : 

With this code

```ts
// /pages/_middleware.ts
import { Settings } from 'luxon'

export async function middleware() {
    Settings.defaultZone = 'utc'
}
```

```ts 
// /pages/api/foo.ts
const handle = async (req: NextApiRequest, res: NextApiResponse): Promise<void> => {
    console.log(Settings.defaultZone)
}

export default handle
```

with next 12.0.7, this will output utc (or at least the corresponding data)
with next 12.0.8, this will output the local timezone (or at least the corresponding data)

So it seems that the "process" (or whatever it is) is not kept/persisted

**Comment by alessandrojcm** [comment]
FWIW I've been having this issue as well very frequently with 12.0.8; virtually every change triggers a full request of the page, renaming the middleware didn't help (I only have one global middleware and it does not have any external imports).

**Comment by ChiefORZ** [comment]
> FWIW I've been having this issue as well very frequently with 12.0.8; virtually every change triggers a full request of the page, renaming the middleware didn't help (I only have one global middleware and it does not have any external imports).

for us the workflow is deleting all the middlewares before starting on features. and to avoid committing the deleted files, weve created a pre-commit check that fails if they get removed 😎

**Comment by alessandrojcm** [comment]
> > FWIW I've been having this issue as well very frequently with 12.0.8; virtually every change triggers a full request of the page, renaming the middleware didn't help (I only have one global middleware and it does not have any external imports).
> 
> for us the workflow is deleting all the middlewares before starting on features. and to avoid committing the deleted files, weve created a pre-commit check that fails if they get removed 😎

Nice! Any chance you could share that config please?

**Comment by pqt** [comment]
> for us the workflow is deleting all the middlewares before starting on features. and to avoid committing the deleted files, weve created a pre-commit check that fails if they get removed 😎

See, for me, the workflow is that I'm pinned to v12.0.4 – but subscribed to updates in this issue – until this is closed from a merged PR that I can safely update. 😂

**Comment by balazsorban44** [comment]
Hi, this has been fixed on `latest` (as of writing `12.0.9`) please upgrade and give it a try!

**Comment by switz** [comment]
Still seeing this issue on `12.0.9`

<img width="448" alt="image" src="https://user-images.githubusercontent.com/990166/151300454-b5734a55-a644-41ff-90d9-01b41acc6ccf.png">

**Comment by bmichotte** [comment]
<img width="979" alt="image" src="https://user-images.githubusercontent.com/235510/151308827-10571633-9cb8-401b-b2bc-860cb74b243a.png">

Same here

**Comment by klapec** [comment]
Yup, definitely not fixed on 12.0.9.
Here's the stacktrace that I'm getting:
```
TypeError: Cannot read properties of undefined (reading 'call')
    at Object.options.factory (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643268216530:685:31)
    at __webpack_require__ (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643268216530:37:33)
    at fn (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643268216530:354:21)
    at eval (webpack-internal:///./app.config.ts:30:68)
```
As soon as I remove any module imports - the problem is gone.

**Comment by simontaisne** [comment]
Same here on `12.0.9` and without any import outside of `next/server` :(

```
TypeError: Cannot read properties of undefined (reading 'call')
    at Object.options.factory (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643273407380:690:31)
    at __webpack_require__ (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643273407380:37:33)
    at fn (http://localhost:3000/_next/static/chunks/webpack.js?ts=1643273407380:359:21)
    at eval (webpack-internal:///./src/hooks/useApollo.ts:15:40)
```

**Comment by bmichotte** [comment]
Hello @balazsorban44, I just created a quick reproduction repository : https://github.com/bmichotte/next-issue-31827

With that project cloned, you have to

1.  `yarn install`
2. `yarn dev`
3. open http://localhost:3000/ (and enjoy an _Hello world_ :P)
4. wait a few seconds, then uncomment the `console.log` in the `api/v1/hello.ts` file
5. save that file
6. reproduce 4-5 with commenting/uncommenting the console.log, you should see the full reload

**Comment by iksent** [comment]
So now we need to close this annoying notice "About to perform a pull refresh" in addition to page refreshing 😆

**Comment by seonghyeonkimm** [comment]
next: 12.0.9
node: 16.13.0

I also ran into this warning dialog. And I think it is related to @sentry/nextjs withSentryConfig calls in next.config.js.

<img width="354" alt="스크린샷 2022-02-01 오전 12 58 01" src="https://user-images.githubusercontent.com/13966404/151827752-7523b965-6625-4d60-b443-fd42bbf161c9.png">

**Comment by jescalan** [comment]
Hi friends! Just want to say that we are aware of this issue on the next.js core team and are working towards getting a fix out. Really appreciate your patience here, and so sorry to everyone who this has inconvenienced!

**Comment by ijjk** [comment]
Hi, this has been updated in `v12.0.11-canary.1` of Next.js, please update and give it a try!

**Comment by bmichotte** [comment]
@ijjk I can confirm that I don't see the complete refresh in the two projects I tested with that version !

**Comment by switz** [comment]
Can also confirm this is fixed. Thank you so so so much!!

**Comment by GandharvJaggi** [comment]
HI, we are using a custom server.js . Upgrading to `v12.0.11-canary.1` did not fix the problem

**Comment by ijjk** [comment]
@GandharvJaggi if you are not using middleware then this sounds unrelated to your issue, can you open a separate issue with a reproduction to allow investigation?

**Comment by pqt** [comment]
@jescalan & @ijjk since serving the public* is an often thankless task, I just want to say thank you **so much** for this, it has really been a thorn in the side and this is probably the highlight to the day.

Just wanted to share that, pass it along to anyone involved, pats on the back for all.

**Comment by sebastienbarre** [comment]
I tried v12.0.11-canary.10 and I still get a lot of those but in the context of using MDX files. Would you mind considering an option to disable this big alert box, and reload automatically? Or is there an easy to do it already? Thank you

**Comment by sebastienbarre** [comment]
Thanks for your time @ijjk. I created #34212, with a video capture and a repo.
